src/util/file_watch.c

```diff
@@ -0,0 +1,331 @@
+            DEBUG(SSSDBG_OP_FAILURE, "Cannot create poll timer\n");
+            return ret;
+        }
+    }
+
+    return EOK;
+}
+
+
+static int watch_file(TALLOC_CTX *mem_ctx,
+                      struct file_watch_ctx *fw_ctx)
+{
+    int ret = EOK;
+    bool use_inotify;
+
+    use_inotify = fw_ctx->use_inotify;
+    if (use_inotify) {
+        ret = try_inotify(fw_ctx);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_MINOR_FAILURE, "Falling back to polling\n");
+            use_inotify = false;
+        }
+    }
+
+    if (!use_inotify) {
+        ret = fw_watch_file_fallback(mem_ctx, fw_ctx);
+    }
+
+    return ret;
+}
+
+
+static void missing_file(struct tevent_context *ev,
+                                struct tevent_timer *te,
+                                struct timeval tv, void *data)
+{
+    int ret;
+    struct file_watch_ctx *fw_ctx = talloc_get_type(data, struct file_watch_ctx);
+
+    ret = watch_file(fw_ctx, fw_ctx);
+    if (ret == EOK) {
+        fw_ctx->fn(fw_ctx->filename, fw_ctx->fn_arg);
+    } else if (ret == ENOENT) {
+        DEBUG(SSSDBG_MINOR_FAILURE,
+              "%s missing. Waiting for it to appear.\n",
+              fw_ctx->filename);
+        tv = tevent_timeval_current_ofs(MISSING_FILE_POLL_TIME, 0);
+        te = tevent_add_timer(fw_ctx->ev, fw_ctx, tv, missing_file, fw_ctx);
+        if (te == NULL) {
+            DEBUG(SSSDBG_FATAL_FAILURE,
+                  "tevent_add_timer failed. %s will be ignored.\n",
+                  fw_ctx->filename);
+        }
+    } else {
+        DEBUG(SSSDBG_FATAL_FAILURE,
+              "watch_file failed. %s will be ignored.\n",
+              fw_ctx->filename);
+    }
+}
+
+
+struct file_watch_ctx *fw_watch_file(TALLOC_CTX *mem_ctx,
+                                     struct tevent_context *ev,
+                                     const char *filename,
+                                     bool use_inotify,
+                                     fw_fn fn,
+                                     void *fn_arg)
+{
+    int ret;
+    struct timeval tv;
+    struct tevent_timer *te;
+    struct file_watch_ctx *fw_ctx;
+
+    if (ev == NULL || filename == NULL || fn == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Invalid parameter\n");
+        return NULL;
+    }
+
+    fw_ctx = talloc_zero(mem_ctx, struct file_watch_ctx);
+    if (fw_ctx == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Failed to allocate the context\n");
+        return NULL;
+    }
+
+    fw_ctx->ev = ev;
+    fw_ctx->use_inotify = use_inotify;
+    fw_ctx->fn = fn;
+    fw_ctx->fn_arg = fn_arg;
+    fw_ctx->filename = talloc_strdup(fw_ctx, filename);
+    if (fw_ctx->filename == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "talloc_strdup() failed\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    /* Watch for changes to the requested file */
+    ret = watch_file(fw_ctx, fw_ctx);
+    if (ret == ENOENT) {
+        tv = tevent_timeval_current_ofs(MISSING_FILE_POLL_TIME, 0);
```

Comment:
Just call `missing_file()` to avoid copy&paste a bit?

Reply:
Just wonder why this was done this complex way if this is not needed.

Reply:
This code was just moved and proved working. The less modifications I do, the less chances that I introduce a bug. That's why I try to maximize the copy&paste.

Reply:
Let's take the opportunity to improve evident things. :)

Reply:
Is this really an issue to check a file twice without delay?
This would simplify code a bit.

Reply:
I simplified a bit this piece of code. However, this introduces a slight change in behavior because of what seems to be a change in behavior in inotify. This is described in the commit message, which I copy here below:

> This code is needlessly repeated in fw_watch_file() and missing_file(), and it's enough to simply call missing_file(). But since now this function is used even before knowing whether the file exists, it is better to rename it to set_file_watching().
> 
> A side effect of this is that watch_file() (which ends up calling inotify_add_watch()) returns EOK even when the file is missing, and thus the callback is invoked once. Considering that the flag IN_CREATE was always passed to inotify_add_watch(), I think this could have been hidden by the old code.
>
> Polling for the file creation is no longuer used when inotify is used. But because it is still required when inotify fails to initialize (not installed) or use_inotify is set to false, this code cannot be removed.

Reply:
> thus the callback is invoked once

Will this trigger double `resolv_reread_configuration()` at startup?


>  I think this could have been hidden by the old code.

What do you mean?

> Polling for the file creation is no longuer used when inotify is used.

Was polling really used along with inotify before?
`monitor_config_file_fallback()` was only executed if `(use_inotify == false)` (as it kept).

Reply:
> > thus the callback is invoked once
> 
> Will this trigger double `resolv_reread_configuration()` at startup?

Could happen. Is that a problem? I see it does things with c-ares, but I'm not fully aware about what that implies.


> > I think this could have been hidden by the old code.
> 
> What do you mean?

Argh. The right sentence should have been "Considering that the flag IN_CREATE was always passed to `inotify_add_watch()`, I think it is possible the behavior changed at some point in time and this could have been hidden by the old code.

In short, the `IN_CREATE` flag controls whether the callback will be called when the file is created. As a side effect, the call to `inotify_add_watch()` will succeed even if the file does nor exist. This flag was already in use, but the code was written to detect by itself the file creation. The only reasons I can imagine for this is that the inotify failed to detect the file creation at some time and later it was fixed, or that `IN_CREATE`'s behavior was extended to cover this case. When this change happened, maybe the old code hid this somehow (there was some change but was lost in the logs and nobody noticed because the final behavior was good).

> > Polling for the file creation is no longuer used when inotify is used.
> 
> Was polling really used along with inotify before? IIUC, `monitor_config_file_fallback()` was only executed if `(use_inotify == false)` (like now).

Never at the same time, but the polling was launched when `inotify_add_watch()` failed because the file didn't exist.
Don' t forget `use_inotify` would be set to `false` if `try_inotify()` returned an error.

Reply:
>>  Will this trigger double resolv_reread_configuration() at startup?
> Could happen. Is that a problem? I see it does things with c-ares, but I'm not fully aware about what that implies.

I think it's possible (and better) to avoid this.

> This flag was already in use, but the code was written to detect by itself the file creation.
...
> when snotify_watch() failed because the file didn't exist

How would this be possible?
Reading man page, it's unclear to me if `inotify_add_watch()` would ever fail in this case. The only relevant error code is `ENOENT` that talks about missing dir component (not file), and, btw, this error code doesn't make `snotify_watch()` to fail.

Reply:
> I think it's possible (and better) to avoid this.

I mean, if we want reusable code, it's better to avoid "false" events as it could potentially trigger heavy processing.

Reply:
> > > > Will this trigger double resolv_reread_configuration() at startup?
> > > > Could happen. Is that a problem? I see it does things with c-ares, but I'm not fully aware about what that implies.
> > 
> > I think it's possible (and better) to avoid this.
> 
> This is corrected now. I removed the extra security that was actually useless and inotify or polling will correctly detect any change.

Did you push the latest version?

Looks like callback still invoked immediately if file exist.

Reply:
Would it be possible to use `te` to distinguish context `set_file_watching()` is called from and skip callback during initial setup?

Reply:
> > > > > Will this trigger double resolv_reread_configuration() at startup?
> > > > > Could happen. Is that a problem? I see it does things with c-ares, but I'm not fully aware about what that implies.
> > > 
> > > 
> > > I think it's possible (and better) to avoid this.
> > 
> > 
> > This is corrected now. I removed the extra security that was actually useless and inotify or polling will correctly detect any change.
> 
> Did you push the latest version?
> 
> Looks like callback still invoked immediately if file exist.

Yes, I did push the changes. The goal of those changes was to avoid a double invocation of the callback. Now it is invoked only once.

Reply:
> 

Do we really want to avoid the callback being called at start up? The file needs to be read once at the beginning, and this is when it is done. The only "problem" occurs when the file doesn't exist but the callback is still called (and will be called again when the file exists). Maybe we should focus on this case, for instance, not invoking the callback if the file doesn't exist.

Reply:
Well, I found the problem. The code I removed was very similar to the code in the `fw_watch_file_fallback()` function but not identical. There was one single line which was different and it invoked the callback. That's what makes the callback to be called when `watch_file()` succeeds (always with inotify).

There are two solutions:
1. Go back to the old code which was written specifically to avoid this situation, or
2. In `set_file_watching()` check for the existence of the file before calling the callback when `watch_file()` returns `EOK`.

I think second is better. What's your opinion?

Reply:
Well, my idea was to avoid double execution of `resolv_reread_configuration()` at startup.
But I realized that it isn't executed at startup unconditionally anymore, so immediate callback isn't an issue (but would be good to document this behavior in `file_watch.h`)
But I now wonder if `resolv_reread_configuration()`/`check_if_online()` is still required at startup even with missing resolv.conf (probably not)...

Reply:
Well, with the latest modifications, the callback is not called if the file is missing. So this is not a problem any more.

Reply:
I also improved the test to check this case.

Reply:
Ok.

Could you please also add a comment to `fw_watch_file()` function in `src/util/file_watch.h` file, describing logic in general (`use_inotify`/polling) and this detail (immediate cb invocation in case file exist) in particular?

Reply:
Done

Reply:
Thank you.